debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/ehcache/EhcacheLogMinerTransactionCache.java

```diff
@@ -0,0 +1,187 @@
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            return consumer.apply(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void transactions(Consumer<Stream<EhcacheTransaction>> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void eventKeys(Consumer<Stream<String>> consumer) {
+        try (var stream = StreamSupport.stream(eventCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getKey));
+        }
+    }
+
+    @Override
+    public Iterator<LogMinerEvent> eventsIterator(EhcacheTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId())
+                .stream()
+                .map(eventId -> getTransactionEvent(transaction, eventId))
+                .iterator();
+    }
+
+    @Override
+    public LogMinerEvent getTransactionEvent(EhcacheTransaction transaction, int eventKey) {
+        return eventCache.get(transaction.getEventId(eventKey));
+    }
+
+    @Override
+    public EhcacheTransaction getAndRemoveTransaction(String transactionId) {
+        final EhcacheTransaction transaction = getTransaction(transactionId);
+        if (transaction != null) {
+            transactionCache.remove(transactionId);
+        }
+        return transaction;
+    }
+
+    @Override
+    public void addTransactionEvent(EhcacheTransaction transaction, int eventKey, LogMinerEvent event) {
+        eventCache.put(transaction.getEventId(eventKey), event);
+        eventIdsByTransactionId.get(transaction.getTransactionId()).add(eventKey);
+    }
+
+    @Override
+    public void removeTransactionEvents(EhcacheTransaction transaction) {
+        eventIdsByTransactionId.get(transaction.getTransactionId())
+                .descendingSet()
+                .stream()
+                .map(transaction::getEventId)
+                .forEach(eventCache::remove);
+
+        eventIdsByTransactionId.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean removeTransactionEventWithRowId(EhcacheTransaction transaction, String rowId) {
+        final TreeSet<Integer> eventIds = eventIdsByTransactionId.get(transaction.getTransactionId());
+        return eventIds.descendingSet().stream()
+                .map(i -> Map.entry(i, transaction.getEventId(i)))
+                .filter(entry -> {
+                    final LogMinerEvent event = eventCache.get(entry.getValue());
+                    return event != null && event.getRowId().equals(rowId);
+                })
+                .findFirst()
+                .map(entry -> {
+                    eventCache.remove(entry.getValue());
+                    eventIds.remove(entry.getKey());
```

Comment:
We're removing entries from the TreeSet while streaming over it.  Should copy the set and iterate over that.

Reply:
I understand the concern but given this is in `findFirst()` and the stream breaks after the removal, I'm not sure its an issue necessarily and I'd prefer to avoid a copy if it's not necessary.

Reply:
Agreed about findFirst there.  Was thinking it was part of the stream when reading it.